processing/src/main/java/io/druid/query/search/search/AutoStrategy.java

```diff
@@ -0,0 +1,56 @@
+package io.druid.query.search.search;
+
+import com.metamx.emitter.EmittingLogger;
+import io.druid.collections.bitmap.ImmutableBitmap;
+import io.druid.query.dimension.DimensionSpec;
+import io.druid.segment.QueryableIndex;
+import io.druid.segment.Segment;
+
+import java.util.List;
+
+public class AutoStrategy extends SearchStrategy
+{
+  public static final String NAME = "auto";
+
+  private static final EmittingLogger log = new EmittingLogger(AutoStrategy.class);
+
+  public AutoStrategy(SearchQuery query)
+  {
+    super(query);
+  }
+
+  @Override
+  public List<SearchQueryExecutor> getExecutionPlan(SearchQuery query, Segment segment)
+  {
+    final QueryableIndex index = segment.asQueryableIndex();
+
+    if (index != null) {
+      final ImmutableBitmap timeFilteredBitmap = IndexOnlyStrategy.makeTimeFilteredBitmap(index,
+                                                                                          segment,
+                                                                                          filter,
+                                                                                          interval);
+      final Iterable<DimensionSpec> dimsToSearch = getDimsToSearch(index.getAvailableDimensions(),
+                                                                   query.getDimensions());
+
+      // Choose a search query execution strategy depending on the query.
+      // Index-only strategy is selected when
+      // 1) there is no filter,
+      // 2) the total cardinality is very low, or
+      // 3) the filter has a very low selectivity.
```

Comment:
And for ```timeFilteredBitmap```, making it is expensive because it requires a lot of unions of bitmaps for filters.
I have a plan to avoid unions by estimating filter selectivity. I already have some benchmark result which shows nearly optimal performance. However, this work needs some new codes and refactoring, so I think it would be better to proceed in another issue. I'll make a PR soon.